File: examples/pydantic/weather-app/run_mcp.py
import warnings
from typing import Any
from automcp.adapters.pydantic import create_pydantic_adapter
from pydantic import BaseModel
from mcp.server.fastmcp import FastMCP

# Create MCP server
mcp = FastMCP("MCP Server")

# Suppress warnings that might interfere with STDIO transport
warnings.filterwarnings("ignore")

# You'll need to replace these imports with your actual pydantic objects
from main import weather_agent


# Define the input schema for your pydantic
class InputSchema(BaseModel):
    # Replace these with your actual input parameters
    query: str
    # Add more parameters as needed


name = "Weather_Agent"
description = "A weather agent that can help you get the weather of a city"

# Create an adapter for pydantic
mcp_pydantic_agent = create_pydantic_adapter(
    agent_instance=weather_agent,  # Replace with your actual Pydantic agent instance
    name=name,
    description=description,
    input_schema=InputSchema,
)


mcp.add_tool(mcp_pydantic_agent, name=name, description=description)


# Server entrypoints
def serve_sse():
    mcp.run(transport="sse")


def serve_stdio():
    # Redirect stderr to suppress warnings that bypass the filters
    import os
    import sys

    class NullWriter:
        def write(self, *args, **kwargs):
            pass

        def flush(self, *args, **kwargs):
            pass

    # Save the original stderr
    original_stderr = sys.stderr

    # Replace stderr with our null writer to prevent warnings from corrupting STDIO
    sys.stderr = NullWriter()

    # Set environment variable to ignore Python warnings
    os.environ["PYTHONWARNINGS"] = "ignore"

    try:
        mcp.run(transport="stdio")
    finally:
        # Restore stderr for normal operation
        sys.stderr = original_stderr


if __name__ == "__main__":
    import sys

    if len(sys.argv) > 1 and sys.argv[1] == "sse":
        serve_sse()
    else:
        serve_stdio()

File: examples/pydantic/weather-app/run_mcp.ts
import { Server } from '@modelcontextprotocol/sdk/server/index.js';
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import { CallToolRequestSchema, ListToolsRequestSchema } from '@modelcontextprotocol/sdk/types.js';
import { z } from 'zod';

// Import our weather agent
import { WeatherAgent } from './main.js';

// Create MCP server
const server = new Server(
  {
    name: 'weather-mcp-server',
    version: '0.1.0',
  },
  {
    capabilities: {
      tools: {},
    },
  }
);

// Define the input schema using Zod (TypeScript equivalent of Pydantic BaseModel)
const InputSchema = z.object({
  query: z.string().describe('The weather query to process'),
});

type InputType = z.infer<typeof InputSchema>;

// Create the weather agent instance
const weatherAgent = new WeatherAgent(
  'openai:gpt-4o-mini',
  'Be concise, reply with one sentence. ' +
  'Use the getLatLng tool to get the latitude and longitude of the locations, ' +
  'then use the getWeather tool to get the weather.',
  2
);

const toolName = 'weather_agent';
const toolDescription = 'A weather agent that can help you get the weather of a city';

// Add tool handlers
server.setRequestHandler(ListToolsRequestSchema, async () => {
  return {
    tools: [
      {
        name: toolName,
        description: toolDescription,
        inputSchema: {
          type: 'object',
          properties: {
            query: {
              type: 'string',
              description: 'The weather query to process',
            },
          },
          required: ['query'],
        },
      },
    ],
  };
});

server.setRequestHandler(CallToolRequestSchema, async (request) => {
  const { name, arguments: args } = request.params;

  if (name === toolName) {
    try {
      // Validate input using Zod schema
      const validatedInput = InputSchema.parse(args);
      
      // Call the weather agent
      const result = await weatherAgent.runSync(validatedInput.query);
      
      return {
        content: [
          {
            type: 'text',
            text: result.data,
          },
        ],
      };
    } catch (error) {
      return {
        content: [
          {
            type: 'text',
            text: `Error: ${error instanceof Error ? error.message : 'Unknown error'}`,
          },
        ],
        isError: true,
      };
    }
  }

  throw new Error(`Unknown tool: ${name}`);
});

// Server entrypoints
async function serveStdio(): Promise<void> {
  // Suppress console.error to prevent warnings from corrupting STDIO
  const originalConsoleError = console.error;
  console.error = () => {}; // Ignore all error outputs

  try {
    const transport = new StdioServerTransport();
    await server.connect(transport);
    
    // Keep the server running
    await new Promise(() => {}); // Run indefinitely
  } finally {
    // Restore console.error for normal operation
    console.error = originalConsoleError;
  }
}

async function serveSSE(): Promise<void> {
  // Note: SSE transport would require additional setup
  // For now, we'll just log that SSE is not implemented
  console.error('SSE transport not implemented in this example');
  console.error('Use STDIO transport instead');
  process.exit(1);
}

// Main execution
async function main(): Promise<void> {
  const args = process.argv.slice(2);
  
  if (args.length > 0 && args[0] === 'sse') {
    await serveSSE();
  } else {
    await serveStdio();
  }
}

// Execute if this is the main module
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((error) => {
    console.error('Failed to start server:', error);
    process.exit(1);
  });
}

export { server, weatherAgent }; 